File: core/src/com/broll/gainea/server/core/cards/impl/play/C_HuntMonster.java
package com.broll.gainea.server.core.cards.impl.play;

import com.broll.gainea.server.core.cards.Card;
import com.broll.gainea.server.core.map.Location;
import com.broll.gainea.server.core.objects.BattleObject;
import com.broll.gainea.server.core.utils.LocationUtils;
import com.broll.gainea.server.core.utils.PlayerUtils;

import java.util.List;
import java.util.stream.Collectors;

public class C_HuntMonster extends Card {
    public C_HuntMonster() {
        super(37, "Drachenjagd", "Wählt eine Truppe und greift damit ein beliebiges Monster auf der gleichen Karte an");
    }

    @Override
    public boolean isPlayable() {
        return true;
    }

    @Override
    protected void play() {
        Location from = selectHandler.selectLocation("Wählt eure Angriffstruppe", owner.getControlledLocations());
        List<BattleObject> attackers = PlayerUtils.getUnits(owner, from);
        List<Location> attackLocations = LocationUtils.getWildMonsterLocations(game).stream()
                .filter(it -> it.getContainer().getExpansion() == from.getContainer().getExpansion()).collect(Collectors.toList());
        if(!attackLocations.isEmpty()){
            Location target = selectHandler.selectLocation("Wählt den Angriffsort aus", attackLocations);
            game.getBattleHandler().startBattle(attackers, LocationUtils.getMonsters(target));
        }
    }

}

File: server/src/main/com/broll/gainea/server/core/cards/impl/play/C_HuntMonster.kt
package com.broll.gainea.server.core.cards.impl.play

import com.broll.gainea.server.core.cards.Card
import com.broll.gainea.server.core.cards.EffectType.MOVEMENT
import com.broll.gainea.server.core.utils.getMonsters
import com.broll.gainea.server.core.utils.getUnits
import com.broll.gainea.server.core.utils.getWildMonsterLocations

class C_HuntMonster : Card(
    37,
    MOVEMENT,
    "Drachenjagd",
    "Wählt eine Truppe und greift damit ein beliebiges Monster auf der gleichen Karte an"
) {
    override val isPlayable: Boolean
        get() = true

    override fun play() {
        val from = selectHandler.selectLocation(
            "Wählt eure Angriffstruppe",
            owner.controlledLocations.toList()
        )
        val attackers = owner.getUnits(from)
        val attackLocations = game.getWildMonsterLocations()
            .filter { it.container.expansion === from.container.expansion }
        if (attackLocations.isNotEmpty()) {
            val target = selectHandler.selectLocation("Wählt den Angriffsort aus", attackLocations)
            game.battleHandler.startBattle(attackers, target.getMonsters())
        }
    }
}
